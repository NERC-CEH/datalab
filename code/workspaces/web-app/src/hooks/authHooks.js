import useShallowSelector from './useShallowSelector';
import authSelectors from '../selectors/authSelectors';

const { currentUserId, currentUserPermissions, currentUserTokens } = authSelectors;

export const useCurrentUserId = () => useShallowSelector(currentUserId);
export const useCurrentUserPermissions = () => useShallowSelector(currentUserPermissions);
export const useCurrentUserTokens = () => useShallowSelector(currentUserTokens);
