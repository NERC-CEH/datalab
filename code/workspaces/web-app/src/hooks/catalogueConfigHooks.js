import useShallowSelector from './useShallowSelector';
import catalogueConfigSelectors from '../selectors/catalogueConfigSelectors';

export const useCatalogueConfig = () => useShallowSelector(catalogueConfigSelectors.catalogueConfig);
export const useCatalogueAvailable = () => useShallowSelector(catalogueConfigSelectors.catalogueAvailable);
