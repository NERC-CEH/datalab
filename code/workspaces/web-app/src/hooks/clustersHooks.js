import useShallowSelector from './useShallowSelector';
import clustersSelectors from '../selectors/clustersSelectors';

// eslint-disable-next-line import/prefer-default-export
export const useClusters = () => useShallowSelector(clustersSelectors.clusters);
export const useClustersByType = (type) => {
  const clusters = useClusters();

  if (!type) {
    return clusters;
  }

  return {
    ...clusters,
    value: clusters.value.filter(cluster => cluster.type === type),
  };
};
