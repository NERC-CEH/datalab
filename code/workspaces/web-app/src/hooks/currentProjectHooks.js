import useShallowSelector from './useShallowSelector';
import currentProjectSelectors from '../selectors/currentProjectSelectors';

const { currentProject, currentProjectKey } = currentProjectSelectors;

export const useCurrentProject = () => useShallowSelector(currentProject);
export const useCurrentProjectKey = () => useShallowSelector(currentProjectKey);
