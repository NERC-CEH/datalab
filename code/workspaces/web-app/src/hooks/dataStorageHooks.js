import useShallowSelector from './useShallowSelector';
import dataStorageSelectors from '../selectors/dataStorageSelectors';

// eslint-disable-next-line import/prefer-default-export
export const useDataStorageArray = () => useShallowSelector(dataStorageSelectors.dataStorageArray);
export const useDataStorageForUserInProject = (userId, projectKey) => {
  const dataStores = useDataStorageArray();
  return {
    ...dataStores,
    value: dataStores.value.filter(store => store.projectKey === projectKey && store.users.includes(userId)),
  };
};
