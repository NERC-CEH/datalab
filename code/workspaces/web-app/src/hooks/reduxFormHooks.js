import { useSelector } from 'react-redux';
import reduxFormSelectors from '../selectors/reduxFormSelectors';

// eslint-disable-next-line import/prefer-default-export
export const useReduxFormValue = (formName, fieldName) => useSelector(
  reduxFormSelectors.reduxFormValue(formName, fieldName),
);
