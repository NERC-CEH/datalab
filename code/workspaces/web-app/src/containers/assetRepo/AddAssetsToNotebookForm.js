import React from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm } from 'redux-form';
import { useSelector } from 'react-redux';
import { useReduxFormValue } from '../../hooks/reduxFormHooks';
import { UpdateFormControls, formatAndParseMultiSelect, renderSelectField } from '../../components/common/form/controls';
import AssetMultiSelect from '../../components/common/form/AssetMultiSelect';
import validator from './addAssetsToNotebookValidator';

export const FORM_NAME = 'addAssetsToNotebook';
export const PROJECT_FIELD_NAME = 'project';
export const NOTEBOOK_FIELD_NAME = 'notebook';
export const EXISTING_ASSETS_FIELD_NAME = 'existingAssets';
export const ASSETS_FIELD_NAME = 'assets';

const AddAssetsToNotebookFormContent = ({ handleSubmit, projectOptions, notebookOptions, handleClear }) => {
  const selectedProject = useReduxFormValue(FORM_NAME, PROJECT_FIELD_NAME);
  const selectedNotebook = useReduxFormValue(FORM_NAME, NOTEBOOK_FIELD_NAME);
  const notebooks = useSelector(s => s.stacks.value);

  return (
    <div>
      <form onSubmit={handleSubmit} id={'submit'}>
        <div>
          <Field
            name={PROJECT_FIELD_NAME}
            label={'Project'}
            component={renderSelectField}
            options={projectOptions}
          />
        </div>
        <div>
          <Field
            disabled={!selectedProject}
            name={NOTEBOOK_FIELD_NAME}
            label={'Notebook'}
            component={renderSelectField}
            options={notebookOptions}
          />
        </div>
        {selectedNotebook && <div>
          <Field
            disabled
            placeholder={''}
            name={EXISTING_ASSETS_FIELD_NAME}
            label={'Existing Assets on Notebook'}
            component={AssetMultiSelect}
            projectKey={selectedProject}
            format={formatAndParseMultiSelect}
            parse={formatAndParseMultiSelect}
          />
        </div>}
        <div>
          <Field
            disabled={!(selectedProject && selectedNotebook)}
            name={ASSETS_FIELD_NAME}
            label={'Assets'}
            component={AssetMultiSelect}
            projectKey={selectedProject}
            format={formatAndParseMultiSelect}
            parse={formatAndParseMultiSelect}
          />
        </div>
        <UpdateFormControls onCancel={handleClear(notebooks, selectedProject, selectedNotebook)} cancelButtonText={'Reset'}/>
      </form>
    </div>
  );
};

const options = PropTypes.shape({
  text: PropTypes.string,
  value: PropTypes.string,
});

const formPropTypes = {
  projectOptions: PropTypes.arrayOf(options),
  notebookOptions: PropTypes.arrayOf(options),
};

AddAssetsToNotebookFormContent.propTypes = {
  ...formPropTypes,
  handleSubmit: PropTypes.func.isRequired,
  handleClear: PropTypes.func.isRequired,
};

const AddAssetsToNotebookForm = reduxForm({
  form: FORM_NAME,
  validate: validator,
})(AddAssetsToNotebookFormContent);

AddAssetsToNotebookForm.propTypes = {
  ...formPropTypes,
  onSubmit: PropTypes.func.isRequired,
};

export {
  AddAssetsToNotebookFormContent as PureAddAssetsToNotebookForm,
  AddAssetsToNotebookForm,
};
