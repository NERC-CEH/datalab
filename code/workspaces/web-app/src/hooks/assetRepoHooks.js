import useShallowSelector from './useShallowSelector';
import assetRepoSelectors from '../selectors/assetRepoSelectors';
import sortByName from '../components/common/sortByName';
import { PUBLIC } from '../components/assetRepo/assetVisibilities';

const useAssetRepoUnsorted = () => useShallowSelector(assetRepoSelectors.assetRepo);

export const useAssetRepo = () => {
  const assetRepo = useAssetRepoUnsorted();
  return {
    ...assetRepo,
    value:
    {
      ...assetRepo.value,
      assets: sortByName(assetRepo.value.assets),
    },
  };
};

const assetVisible = (projectKey, asset) => asset.fileLocation
  && (asset.visible === PUBLIC
  || (asset.projects && asset.projects.map(project => project.key).includes(projectKey)));

export const useVisibleAssets = (projectKey) => {
  const assetRepo = useAssetRepoUnsorted();
  const visibleAssets = assetRepo.value.assets
    ? assetRepo.value.assets.filter(asset => assetVisible(projectKey, asset))
    : [];
  return {
    ...assetRepo,
    value:
    {
      ...assetRepo.value,
      assets: sortByName(visibleAssets),
    },
  };
};
