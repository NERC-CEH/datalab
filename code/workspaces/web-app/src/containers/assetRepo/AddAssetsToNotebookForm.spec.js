import React from 'react';
import { render, fireEvent, configure } from '@testing-library/react';
import { change, reset } from 'redux-form';
import * as redux from 'react-redux';
import { PureAddAssetsToNotebookForm, FORM_NAME, EXISTING_ASSETS_FIELD_NAME } from './AddAssetsToNotebookForm';
import * as reduxFormHooks from '../../hooks/reduxFormHooks';

jest.mock('redux-form', () => ({
  reduxForm: jest.fn(() => jest.fn()),
  Field: props => (<div>{`Field: ${JSON.stringify(props)}`}</div>),
  change: jest.fn(),
  reset: jest.fn(),
}));

describe('AddAssetsToNotebookForm', () => {
  const dispatchMock = jest.fn().mockName('dispatch');

  beforeEach(() => {
    jest.clearAllMocks();
    configure({ testIdAttribute: 'id' });
    jest.mock('react-redux');
    jest.mock('../../hooks/reduxFormHooks');

    redux.useDispatch = jest.fn().mockReturnValue(dispatchMock);
    redux.useSelector = jest.fn().mockReturnValue([]);
    reduxFormHooks.useReduxFormValue = jest.fn().mockReturnValue('dummyValue');
  });

  const handleSubmit = jest.fn();

  const generateProps = () => ({
    handleSubmit,
    projectOptions: ['project1', 'project2'],
    notebookOptions: ['notebook1', 'notebook2'],
  });

  it('renders correctly when a project and notebook have been selected', () => {
    const props = generateProps();

    reduxFormHooks.useReduxFormValue = jest.fn()
      .mockReturnValueOnce('project1')
      .mockReturnValueOnce('notebook1');

    const wrapper = render(<PureAddAssetsToNotebookForm {...props}/>);

    expect(wrapper.container).toMatchSnapshot();
    expect(wrapper.queryAllByText('Field', { exact: false })).toHaveLength(4);
  });

  it('renders correctly when a no notebook has been selected', () => {
    const props = generateProps();

    reduxFormHooks.useReduxFormValue = jest.fn()
      .mockReturnValueOnce('project1')
      .mockReturnValueOnce();

    const wrapper = render(<PureAddAssetsToNotebookForm {...props}/>);

    expect(wrapper.container).toMatchSnapshot();
    expect(wrapper.queryAllByText('Field', { exact: false })).toHaveLength(3);
  });

  it('dispatches reset when form is cleared', () => {
    const props = generateProps();
    const resetDispatch = jest.fn();
    reset.mockReturnValueOnce(resetDispatch);

    const wrapper = render(<PureAddAssetsToNotebookForm {...props}/>);

    fireEvent.click(wrapper.getByText('Cancel'));
    expect(dispatchMock).toHaveBeenCalledTimes(1);
    expect(dispatchMock).toHaveBeenCalledWith(resetDispatch);
    expect(reset).toHaveBeenCalledTimes(1);
    expect(reset).toHaveBeenCalledWith(FORM_NAME);
  });

  it('calls submit function when the form is submitted', () => {
    const props = generateProps();

    const wrapper = render(<PureAddAssetsToNotebookForm {...props}/>);

    expect(handleSubmit).toHaveBeenCalledTimes(0);
    fireEvent.submit(wrapper.getByTestId('submit'));
    expect(handleSubmit).toHaveBeenCalledTimes(1);
  });

  it('displays existing assets when a notebook is selected', () => {
    const props = generateProps();
    reduxFormHooks.useReduxFormValue = jest.fn()
      .mockReturnValueOnce('project1')
      .mockReturnValueOnce('notebook1');
    const changeDispatch = jest.fn();
    change.mockReturnValueOnce(changeDispatch);

    const assets = ['existing asset1', 'existing asset2'];
    const stacks = [
      {
        projectKey: 'otherProject',
        name: 'notebook1',
        assets: ['not this asset'],
      },
      {
        projectKey: 'project1',
        name: 'notebook2',
        assets: ['or this asset'],
      },
      {
        projectKey: 'project1',
        name: 'notebook1',
        assets,
      },
    ];
    redux.useSelector = jest.fn().mockReturnValue(stacks);

    render(<PureAddAssetsToNotebookForm {...props}/>);

    expect(dispatchMock).toHaveBeenCalledTimes(1);
    expect(dispatchMock).toHaveBeenCalledWith(changeDispatch);
    expect(change).toHaveBeenCalledTimes(1);
    expect(change).toHaveBeenCalledWith(FORM_NAME, EXISTING_ASSETS_FIELD_NAME, assets);
  });
});
