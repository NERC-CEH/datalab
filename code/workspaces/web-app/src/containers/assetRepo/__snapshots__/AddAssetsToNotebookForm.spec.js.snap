// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAssetsToNotebookForm renders correctly when a no notebook has been selected 1`] = `
<div>
  <div>
    <form
      id="submit"
    >
      <div>
        <div>
          Field: {"name":"project","label":"Project","options":[{"text":"project1","value":"project1"},{"text":"project2","value":"project2"}]}
        </div>
      </div>
      <div>
        <div>
          Field: {"disabled":false,"name":"notebook","label":"Notebook","options":[{"text":"notebook1","value":"notebook1"},{"text":"notebook2","value":"notebook2"}]}
        </div>
      </div>
      <div>
        <div>
          Field: {"disabled":true,"name":"assets","label":"Assets","projectKey":"project1"}
        </div>
      </div>
      <div
        class="makeStyles-fromControlContainer-56"
      >
        <span
          class="makeStyles-button-57"
          title=""
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Apply
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </span>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-button-57 MuiButton-outlinedSecondary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Reset
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`AddAssetsToNotebookForm renders correctly when a project and notebook have been selected 1`] = `
<div>
  <div>
    <form
      id="submit"
    >
      <div>
        <div>
          Field: {"name":"project","label":"Project","options":[{"text":"project1","value":"project1"},{"text":"project2","value":"project2"}]}
        </div>
      </div>
      <div>
        <div>
          Field: {"disabled":false,"name":"notebook","label":"Notebook","options":[{"text":"notebook1","value":"notebook1"},{"text":"notebook2","value":"notebook2"}]}
        </div>
      </div>
      <div>
        <div>
          Field: {"disabled":true,"placeholder":"","name":"existingAssets","label":"Existing Assets on Notebook","projectKey":"project1"}
        </div>
      </div>
      <div>
        <div>
          Field: {"disabled":false,"name":"assets","label":"Assets","projectKey":"project1"}
        </div>
      </div>
      <div
        class="makeStyles-fromControlContainer-1"
      >
        <span
          class="makeStyles-button-2"
          title=""
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Apply
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </span>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-button-2 MuiButton-outlinedSecondary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Reset
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </div>
</div>
`;
